.github/workflows/get_layer_info_for_nf_imgs.yml

```diff
@@ -16,29 +16,86 @@ jobs:
+              error_code=$(echo "$response" | jq -r '.errors[0].code // empty')
+
+              if [ "$error_code" == "TOOMANYREQUESTS" ]; then
+                echo "Line $LINENO: Rate limit exceeded, waiting for $wait_time seconds before retrying..." >&2
+                sleep $wait_time
+                wait_time=$((wait_time * 2))  # Exponential backoff
+              else
+                echo "$response"
+                return
+              fi
+            done
+
+            echo "Line $LINENO: Failed to fetch manifest after $retries attempts due to rate limiting." >&2
+            return 1
+          }
+
           while IFS= read -r image_url; do
             # Ignore lines that start with #
               if [[ "$image_url" == \#* ]]; then
-                echo "Skipping this line -- $image_url"
+              echo "Line_$LINENO: Skipping this line -- $image_url"
                 continue
               fi
+
             # Strip the first * if a line starts with *
               if [[ "$image_url" == \** ]]; then
-                echo "Stripping * from this line -- $image_url"
+              echo "Line_$LINENO: Stripping * from this line -- $image_url"
                 image_url="${image_url:1}"
               fi
-              manifest_url=$(echo ${image_url} | sed 's|public\.ecr\.aws/\(.*\):\(.*\)|https://public.ecr.aws/v2/\1/manifests/\2|')
-              tag_name=$(echo ${image_url} | sed 's|\(.*\):\(.*\)|\2|')
-              echo $manifest_url, $tag_name
-              response=$(curl -s -H "Authorization: Bearer $TOKEN" $manifest_url | jq "[.layers[].digest]")
-              layer_json=$(echo "$layer_json" | jq --arg key "$image_url" --argjson extracted "$response" '. + { ($key): $extracted }')
+            echo "Line $LINENO: started reading lines from script"
+            manifest_url=$(echo "${image_url}" | sed 's|public\.ecr\.aws/\(.*\):\(.*\)|https://public.ecr.aws/v2/\1/manifests/\2|')
+            tag_name=$(echo "${image_url}" | sed 's|\(.*\):\(.*\)|\2|')
+            echo "$LINENO: manifest_url = $manifest_url, tag_name = $tag_name"
+            # Fetch the manifest with retry using exponential backoff
+            response=$(fetch_manifest "$manifest_url" | jq "[.layers[].digest]|.[-1]")
+
+            if [ $? -ne 0 ] || [ -z "$response" ]; then
+              echo "Line $LINENO: Failed to retrieve valid response from manifest_url -- $manifest_url" >&2
+              continue
+            fi
+
+            # Extracting current value of `image_url` if the key exists, otherwise return `{}`
+            current_value=$(echo "$layer_json" | jq -e ".[\"$image_url\"] // empty" || echo "{}")
+
+            # Eliminating older values if there are more than `max_version_count` and return all the remaining values.
+            current_value=$(echo "$current_value" | jq --argjson k "$max_version_count" 'to_entries | sort_by(.key) | reverse | if length > $k then .[:$k] else . end | from_entries')
```

Comment:
this step doesn't use the timestamps to remove old values, it assumes the entries are in descending order, right? could you mention that in the comment

Reply:
It doesn't assume the fields to be in desc order, but it actually sorts the fields in descending order, not based on timestamps but using string comparison.